Expose 3 new methods with text metrics in RenderParagraph

pulyaevskiy opened the pull request:
## Description

Exposes `preferredLineHeight`, `getFullHeightForCaret` and `getLineBoundary` on `RenderParagraph`.

## Use case

I've created a [rich text editor](https://github.com/memspace/zefyr) for Flutter and currently work on cleaning up and refactoring its rendering layer.

In the editor I'm taking advantage of the built-in `RenderParagraph` to represent a single paragraph in a document.
Since I'm using it in an editable context I need it to provide a few more metrics related to handling of editing cursor:

* `preferredLineHeight`
* `getFullHeightForCaret`
* `getLineBoundary`

These methods would be proxying to the underlying `TextPainter` instance. Right now I have implemented a workaround for `preferredLineHeight` but it doesn't seem like I can create one for `getFullHeightForCaret` or `getLineBoundary`.

I noticed that `RenderParagraph` already exposes some editing-specific methods (`getOffsetForCaret` and `getBoxesForSelection`) so it doesn't seem like it'd go against certain design goals for this class.

## Related Issues

* Fixes https://github.com/flutter/flutter/issues/65069

## Tests

Added the following tests in `packages/flutter/test/rendering/paragraph_test.dart`:

* `preferredLineHeight control test`
* `getFullHeightForCaret control test`
* `getLineBoundary control test`


## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

pulyaevskiy committed:
Expose more caret API methods on RenderParagraph

pulyaevskiy committed:
Added more tests

pulyaevskiy commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -36,6 +36,42 @@ void main() {
+    layout(paragraph);
+
+    final TextRange range5 = paragraph.getLineBoundary(const TextPosition(offset: 5));
+    expect(range5.textInside(_kText), equals('I polished up that handle so carefullee'));
```
The doc on `TextPainter` says "The newline, if any, is included in the range". However this doesn't seem to be the case... or I missunderstand this statement.

I couldn't find any tests for `TextPainter.getLineBoundary` to see if this is expected.

pulyaevskiy replied:
Oh, interesting. So the web tests failed here with:

```
07:30 +485 ~12 -1: /b/s/w/ir/k/flutter/packages/flutter/test/rendering/paragraph_test.dart: getLineBoundary control test [E]                                                                           
  Expected: 'I polished up that handle so carefullee'
    Actual: 'I polished up that handle so carefullee\n'
              ''
     Which: is different. Both strings start the same, but the actual value also has the following trailing characters: \n
```

Which sounds like implementation for the web platform is more aligned with the doc... ?

Running this test locally on macOS works fine.

I'm probably going to need help determining which behavior is the correct one and whether I should be looking at the TextPainter implementation to fix this inconsistency.

justinmc replied:
I would agree with you, something seems not right there.  Let's see if @GaryQian knows what the behavior should be.

There are some tests in the web part of the engine for this, and they include the newline: https://github.com/flutter/engine/blob/09a5bf7acf44dbb22d68830f5a51ebe822d5e659/lib/web_ui/test/paragraph_test.dart#L926

GaryQian replied:
This may indeed be a bug, I think the newline makes sense here, especially since the docs indicate that it was meant to be included. It isn't your responsibility to fix that though, so you may want to just mark it with a comment and a `TODO: blahblah` to indicate that the result here may be due to a bug. Then, you can duplicate the test and make one of them skip web while having the other skip android+ios. We should file a bug on this.

dnfield commented:
Could you add some more details to the pull request description about the use case(s) for this?

pulyaevskiy commented:
> Could you add some more details to the pull request description about the use case(s) for this?

@dnfield sure, updated the description with my use case.

Hixie commented:
These additions seem fine in general. @GaryQian is the right person to make calls on the details of the API. We definitely want Web and VM to have identical behavior. It might involve changing the documentation or the implementations or both.

For the tests I recommend trying to be more specific in the conditions. The more precise the tests are the less likely it is that one day we will accidentally regress things.

justinmc commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -735,6 +735,11 @@ class RenderParagraph extends RenderBox
+  /// This does not required the layout to be updated.
+  double get preferredLineHeight => _textPainter.preferredLineHeight;
+
+
```
Extra newline here.

justinmc commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -735,6 +735,11 @@ class RenderParagraph extends RenderBox
   }
 
+  /// An estimate of the height of a line in the text. See [TextPainter.preferredLineHeight].
+  /// This does not required the layout to be updated.
```
required => require

justinmc replied:
Or maybe this? "Accessing this value does not cause the layout to be updated."

pulyaevskiy replied:
Removed this method per @GaryQian 's recommendation.

justinmc commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -744,6 +749,15 @@ class RenderParagraph extends RenderBox
 
+  /// Returns the tight bounded height of the glyph at the given [position].
+  ///
+  /// Valid only after [layout].
```
Any of these docs comments that were copied from text_painter.dart should use a macro instead.

Here's an example of how to do that:

https://github.com/flutter/flutter/blob/0bb1e573fce0a32c28d1bd3caf26e53405604f48/packages/flutter/lib/src/painting/text_painter.dart#L335-L337

https://github.com/flutter/flutter/blob/0bb1e573fce0a32c28d1bd3caf26e53405604f48/packages/flutter/lib/src/rendering/paragraph.dart#L269

pulyaevskiy replied:
Thanks for the examples!

justinmc reviewed (commented):
Thanks for opening this!  I'm excited about Flutter being able to support rich text applications like your project.  I'm on board with exposing these methods, but let's wait for Gary to chime in too.

GaryQian commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -735,6 +735,11 @@ class RenderParagraph extends RenderBox
 
+  /// An estimate of the height of a line in the text. See [TextPainter.preferredLineHeight].
+  /// This does not required the layout to be updated.
+  double get preferredLineHeight => _textPainter.preferredLineHeight;
```
If you have a workaround for preferredLineHeight, I would prefer that we hold off on exposing this as official API. The current implementation of this value is sketchy at best and is something I want to eventually rework.

Some problems with this value is that it assumes an alphabetic space as the character to get metrics from. This causes some issues with ideographic text such as Chinese or even Arabic in some fonts. The current implementation also does not account for the differences in first line vs second+ line height, so could produce values that are not useful beyond the first line.

For now, I suggest continuing to use your workaround instead of adding this API that may have to be broken in the future.

pulyaevskiy replied:
Makes sense, thanks! I can keep my workaround for now and remove this method from here.

GaryQian commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -744,6 +749,15 @@ class RenderParagraph extends RenderBox
     return _textPainter.getOffsetForCaret(position, caretPrototype);
   }
 
+  /// Returns the tight bounded height of the glyph at the given [position].
```
tight -> strut

Can you change this in the original docs in TextPainter as well? This doc is no longer accurate.

GaryQian replied:
Going off of the macro comment by Justin, change this in the original doc, and then macro it here.

pulyaevskiy replied:
Updated and added a macro in TextPainter. I haven't included the last line referencing `[layout]` of TextPainter as it seemed specific to the painter class itself (even though RenderParagraph has a similar requirement).

Let me know if you'd like that line to be included in the template, I'll update.

Also changed `[position]` to use backticks, as it seems more inline with the style guide around method parameters.

GaryQian commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -781,6 +795,17 @@ class RenderParagraph extends RenderBox
+  /// The newline, if any, is included in the range.
+  ///
+  /// Valid only after [layout].
+  TextRange getLineBoundary(TextPosition position) {
```
The primary reason this is not exposed is API bloat, since once we add it, it cannot be removed. We should try our best to explore reasonable alternatives before resorting to adding additional API.

In this case, getLineBoundary exposes a few of the text position APIs based on metrics, which we want to eventually move away from to contain more bitdepth.

I think you may be able to workaround this by calling getBoxesForSelection, once with a collapsed selection to get the rect that bounds the current position, and again with a wide selection that encompasses a few lines. Then, you can take the right most and left most boundaries of the boxes that share the same y offset as the first call. Then pass those boundaries to getPositionForOffset.

That should get you something that may work for your purposes. It is a bit of a hassle, but I am hesitant to prematurely expose this particular API.

GaryQian replied:
Let me know if you are able to get a workaround working!

pulyaevskiy replied:
Yep it does seem to work pretty well. Thanks for the suggestion! Since each RenderParagraph in Zefyr represents a single paragraph of text (meaning text between two newlines) here is what I ended up with:

https://github.com/memspace/zefyr/blob/78a5ac11542d39a45cc11cb904416dec0c443da7/packages/zefyr/lib/src/rendering/editable_text_line.dart#L294-L309

```dart
  @override
  TextRange getLineBoundary(TextPosition position) {
    // getOffsetForCaret returns top-left corner of the caret. To find all
    // selection boxes on the same line we shift caret offset by 0.5 of
    // preferredLineHeight so that it's in the middle of the line and filter out
    // boxes which do not include this offset on the Y axis.
    final caret = getOffsetForCaret(position);
    final lineDy = caret.translate(0.0, 0.5 * preferredLineHeight(position)).dy;
    final boxes = getBoxesForSelection(
        TextSelection(baseOffset: 0, extentOffset: node.length - 1));
    final lineBoxes = boxes
        .where((element) => element.top < lineDy && element.bottom > lineDy)
        .toList(growable: false);
    final start = getPositionForOffset(Offset(lineBoxes.first.left, lineDy));
    final end = getPositionForOffset(Offset(lineBoxes.last.right, lineDy));
    return TextRange(start: start.offset, end: end.offset);
  }
```

I'm checking at the vertical middle of the line because I've noticed that emojis, for instance, have larger selection boxes (and they also mess with the line height they are at):
![image](https://user-images.githubusercontent.com/441756/92557187-7db34f80-f220-11ea-89ee-5cc55227e273.png)

Overall it's actually pretty concise and I'm happy with this approach. So I removed `getLineBoundary` from this PR.

It now only exposes `getFullHeightForCaret`.

GaryQian commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -744,6 +749,15 @@ class RenderParagraph extends RenderBox
+  /// Returns the tight bounded height of the glyph at the given [position].
+  ///
+  /// Valid only after [layout].
+  double? getFullHeightForCaret(TextPosition position, Rect caretPrototype) {
```
This should be ok to expose, as the CaretMetrics construct is native to TextPainter.

The name "fullHeight" seems a bit ambiguous in hindsight especially since there is no API for not-full height here. We may want to consider renaming it to something like "rawHeight" here, indicating it is unprocessed and a direct result of the caret computation. The alternative is to just keep the naming as "full" for consistency.

I am leaning towards keeping it as "full" as diverging API is not good. @justinmc any opinions?

justinmc replied:
Is there anything else that refers to "full height" in the codebase?  Not that I could think of or find with a quick search.

If it's really just this method, maybe we should nip it in the bud and switch to "rawHeight" while we have the chance?  If I'm wrong then it definitely makes sense to be consistent and use "fullHeight".

GaryQian reviewed (changes requested):
Left some comments regarding the background/logic around whether to expose API or not.

pulyaevskiy committed:
Addressed PR feedback

pulyaevskiy commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -36,6 +36,19 @@ void main() {
+    const Rect caret = Rect.fromLTWH(0.0, 0.0, 2.0, 20.0);
+
+    final double height5 = paragraph.getFullHeightForCaret(const TextPosition(offset: 5), caret);
+    expect(height5, equals(10.0));
```
Per @Hixie suggestion I updated this test to be more strict in comparing expected value.

pulyaevskiy commented:
@justinmc  @GaryQian thanks for all the feedback! I've updated the PR and removed `preferredLineHeight` and `getLineBoundary`, keeping only `getFullHeightForCaret` with a test. Plus suggested doc changes.

This should be ready for another look.
I'll file a bug for the `TextPainter.getLineBoundary` inconsistency some time tomorrow.

GaryQian reviewed (approved):
LGTM! Thanks for making the changes, and I appreciate your understanding for our efforts in keeping the API clean.

GaryQian commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -744,6 +744,15 @@ class RenderParagraph extends RenderBox
+  double? getFullHeightForCaret(TextPosition position, Rect caretPrototype) {
+    assert(!debugNeedsLayout);
+    _layoutTextWithConstraints(constraints);
+    return _textPainter.getFullHeightForCaret(position, caretPrototype);
```
Oh yeah, one more thing before submitting, caretPrototype here does not do anything useful as in getFullHeightForCaret, it only contributes width, which is then ignored. You should just remove the parameter and pass a zero rect instead in the method body.

pulyaevskiy replied:
Updated.

pulyaevskiy committed:
Removed caretPrototype parameter on getFullHeightForCaret
